fix(popover): Bumping React focus lock to fix autoScroll on autoFocus on popover issue

abhimanyu-singh-uber opened the pull request:
Fixes:- https://codesandbox.io/s/base-web-menu-inside-popover-3vgg1

#### Description
- In Popover, when autoFocus is enabled, with sticky elements it causes an autoScroll on focus as scroll is enabled by default on focus in chrome & firefox.

#### Fix
- Added option in react-focus-lock to preventScroll.
- https://github.com/theKashey/react-focus-lock/pull/186
- https://github.com/theKashey/focus-lock/pull/29
- react-focus-lock@2.7.1 has this option feature so bumping the version to that.


#### Scope
Patch: Bug Fix

williamernest commented:
Thanks for submitting this PR. Can you investigate the failing screenshot tests and resolve those? It looks as if the text in the select is disappearing in certain situations.

lhbrennan commented on src/popover/types.js:
```diff
@@ -104,7 +106,10 @@ export type BasePopoverPropsT = {
-  returnFocus?: boolean,
+  returnFocus?:
+    | boolean
+    | FocusOptions
```
It's a bit confusing why `FocusOptions` can be passed to both `returnFocus`, and to a new, dedicated prop `autoFocusOptions`. It looks like you are mirroring how the API is structure in the source code for `FocusLock`, although I'm not clear on why they've done it that way as well. Is there a good reason for us to do this? If yes, can we add a comment to address the possible confusion?

abhimanyu-singh-uber replied:
yeah, I am just mimicing the prop structure of react-focus-lock here. My reason for doing it here too is to provide the same capability to the baseweb users as the react-focus-lock provides

lhbrennan reviewed (commented):
Overall this looks like a great addition!

Ideally, we also want to demonstrate how this functionality works, both in the docs, and in one of the popover scenarios (`/popover/__tests_/`). Usually it's best to amend one of the existing scenarios, but in this case it might make more sense to create a new scenario file to demonstrate this functionality.

abhimanyu-singh-uber commented on src/popover/popover.js:
```diff
@@ -474,8 +474,9 @@ class PopoverInner extends React.Component<
                 noFocusGuards={false}
                 // see popover-focus-loop.scenario.js for why hover cannot return focus
-                returnFocus={this.props.returnFocus && !this.isHoverTrigger()}
+                returnFocus={!this.isHoverTrigger() && this.props.returnFocus}
```
yes, earlier returnFocus was only `boolean`, but now it can be `boolean | FocusOptions | Function`, by changing the order if `this.props.returnFocus` is a function we would pass a function to `returnFocus`.

abhimanyu-singh-uber commented on src/popover/types.js:
```diff
@@ -83,6 +83,8 @@ export type BasePopoverPropsT = {
   /** Handler for 'Esc' keypress events */
   onFocus?: (e: Event) => mixed,
+  /** Pass FocusOptions to autoFocus */
+  autoFocusOptions?: FocusOptions,
```
I think autoFocusOptions makes more sense imo, because this focusOptions is set & applicable for only autoFocus, but I can change that to focusOptions too to mimic the prop of react-focus-lock

abhimanyu-singh-uber commented on src/combobox/combobox.js:
```diff
@@ -251,6 +251,13 @@ function Combobox<OptionT>(props: PropsT<OptionT>) {
 
+  // React-focus-lock used in Popover used to skip non-tabbable elements (`tabIndex=-1`) elements for focus, we had ListBox with tabIndex to disable focus on
+  // the ListBox, but we can just disable autoFocus (as ListBox / ListItem should not be focusable) (and input is also not autoFocused)
+  const popoverProps = {
```
yeah, we could do that, it is just my personal preference of not doing that. I think if we do that, we cannot override the autoFocus value using restPopoverProps. 
```
const restPopoverProps = { autoFocus: true };
<FocusLock autoFocus={false} {...restPopoverProps} ></FocusLock> 

// In this case, the autoFocus will not be overridden when user passes autoFocus to true for any reason.
```
I see other components (like select) is already doing that (your suggestion) I can do that too.

abhimanyu-singh-uber replied:
But I think we should avoid setting autoFocus to true even if the user passes it as true from above, because it makes the component start behaving wierdly.

Select does the same thing, Let me do that.

Thanks for that advice.

abhimanyu-singh-uber commented:
yes @lhbrennan, I wanted to first check with the team, if they are comfortable with the change before implementing a scenario / test for it. I will add it.

abhimanyu-singh-uber commented:
@williamernest / @lhbrennan I have added the e2e scenarios. & stories, I haven't added the story to documentation-site as the autoscroll only happens when the whole page is scrolled & not a container inside the page. So did not made any sense to me to add it to documentation site.

There is 1 help I would need, that would be get the vrt snapshot images for the new scenario I added in this PR. Can you please help me with that?

Thanks.

abhimanyu-singh-uber commented:
I have fixed the vrt screenshots & merged it into this PR.

A rerun of the ci should pass everything.

Please have a look once you get a chance.

Thanks

abhimanyu-singh-uber commented:
Yay! CIs passed. Got a bit excited as this would be my first PR on baseweb.

abhimanyu-singh-uber commented:
Thanks for the approval, What would be the next steps for merging the PR?

lhbrennan commented:
> Thanks for the approval, What would be the next steps for merging the PR?

Merged, thanks.
